rfcs/records.md

```diff
@@ -0,0 +1,448 @@
+
+Invoking methods with `:` desugars into `getmetatable(obj).method(obj, args)` instead of the usual `obj.method(obj, args)`. This is important because it allows to
+keep the method calls as efficient as possible, as they don't need to check whether the record has a given method as a field.
+
+> TODO: Should we use `__namecall` instead of raw MT access? It seems more consistent, but at the same time `__namecall` today expects a function so it might be best
+> to leave it as is?
+
+> TODO: How do we expose the record keys? Should it be a builtin? Accessible through shape? Not available initially?
+
+### Defining and constructing records
+
+To define a record, you need to create the shape, which you can do using the newly introduced syntax with a context-sensitive `record` keyword.
+
+> TODO: The draft RFC suggests two options for the syntax; only one will be chosen in the final version
+
+Syntax A:
+
+```lua
+record Person = { name: string, age: number }
+
+-- types can be omitted and default to any
+record Point = { x, y }
+```
+
+Syntax B:
+
+```lua
+record Person(name: string, age: number)
+-- types can be omitted and default to any
+record Point(x, y)
+```
+
+This defines `Point` simultaneously as a local variable that corresponds to the shape table, and a type variable that corresponds to the record type.
+
+The resulting shape table automatically is set up to be a valid record shape, but can still be modified by adding methods to it:
+
+```lua
+function Point.__add(l, r)
+    return Point(l.x + r.x, l.y + r.y)
+end
+
+function Point:sum()
+    return self.x + self.y
+end
+
+function Point.newDiagonal(v)
+    return Point(v, v)
+end
+```
+
+Note that `Point` is simply a table and as such it can be used to store static methods as well; as it also serves as a metatable, metafields defined on this table
+will change the behavior of the record values. The shape isn't frozen automatically but can be frozen manually if desired via `table.freeze`.
+
+To create a record, you need to use a record constructor. This is done using call-like syntax:
+
+Syntax A:
+
+```lua
+local person = Person { name = "Bob", age = 42 }
+```
+
+Syntax B:
+
+```lua
+local person = Person("Bob", 42)
+```
+
+The big difference between variants A and B is whether you need to spell out field names at construction time. There's precedents for going either way; some
+languages like F#/C#/Kotlin implement record construction as a function call and when you define a record, you essentially define the record constructor. This
+is beautifully concise, but is a bit more difficult to migrate away from tables, and it's easy to mix up the names. Variant A is more verbose and requires a bit
+more magic at compile time to keep construction efficient.
+
+In variant A, it would make sense to allow omission of any field, in which case it gets replaced with the default of `nil`. A future extension (not part of this RFC) could be made to
```

Comment:
It is equivalent to assigning a field with a non-existent name and as such it'd be an error. I'll clarify that.

Reply:
Hm. With width subtyping incoming, we no longer have a way to say that `A` must have the exact same set of properties as `B`. @asajeffrey you're probably a few steps ahead of me here.

Reply:
Unsealed tables are precise. If https://github.com/Roblox/luau/pull/276 lands, table literals will be unsealed, and so precise.

Reply:
That assumes `Record { x = x }` is the only way to build it, which would prevent you from being able to map some array of tables and apply the record constructor on them. It also means that `f { x = x }` is not going to infer `Callable<<a, b...>({x: a}) -> b...>`. It has to be something else because if you can't write `Record(t)`, then passing `Record` as `f` is an error if it's in the form `f(t)`.

I think the ability to map them is important. One use case might be that you have some data source (for example, parsing some JSON document into a table) and you want to transform this document into an optimized data structure. If you can't apply records like a function call, you have to write things like:

```lua
local function into_city(city: { name: string, population: number}): City
  return City { name = city.name, population = city.population }
end
```

and more for every unique data structure in your JSON document, versus `map(cities, City)` or `City(city)`.